internal/controller/namespacelabel_controller_test.go

```diff
@@ -0,0 +1,175 @@
+package controller
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	labelsv1 "github.com/matanamar10/namespacelabel-operator/api/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+var (
+	ctx       context.Context
+	k8sClient client.Client
+	scheme    *runtime.Scheme
+)
+
+var _ = BeforeSuite(func() {
+	By("Initializing test environment")
+	scheme = runtime.NewScheme()
+	Expect(labelsv1.AddToScheme(scheme)).To(Succeed())
+	Expect(corev1.AddToScheme(scheme)).To(Succeed())
+	k8sClient = fake.NewClientBuilder().WithScheme(scheme).Build()
+	ctx = context.Background()
+})
+
+func createNamespace(name string) {
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+	Expect(k8sClient.Create(ctx, namespace)).To(Succeed())
+}
+
+func deleteNamespace(name string) {
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+	Expect(k8sClient.Delete(ctx, namespace)).To(Succeed())
+}
+
+func deleteAllNamespaceLabels() {
+	namespaceLabelList := &labelsv1.NamespacelabelList{}
+	Expect(k8sClient.List(ctx, namespaceLabelList)).To(Succeed())
+	for _, nl := range namespaceLabelList.Items {
+		Expect(k8sClient.Delete(ctx, &nl)).To(Succeed())
+	}
+}
+
+var _ = Describe("NamespaceLabel Controller", func() {
+	BeforeEach(func() {
+		By("Creating default namespace")
+		createNamespace("default")
+	})
+
+	AfterEach(func() {
+		By("Cleaning up test resources")
+		deleteAllNamespaceLabels()
+		deleteNamespace("default")
+	})
+
+	Context("When reconciling a NamespaceLabel resource", func() {
+		const namespaceName = "default"
+		const resourceName = "test-resource"
+		const secondResourceName = "test-second-resource"
+		namespacedName := types.NamespacedName{Name: resourceName, Namespace: namespaceName}
+
+		It("should successfully create, update, delete labels and delete NamespaceLabels", func() {
```

Comment:
Add tests for the events. See here for an example:

https://github.com/dana-team/cert-external-issuer/blob/main/internal/issuer/controller_test.go